float64 set_end_pos1
float64 set_end_pos2
float64 set_end_pos3